app/invocation/invocation_build_logs_card.tsx

```diff
@@ -1,14 +1,137 @@
 import { TerminalComponent } from "../terminal/terminal";
+import { BuildBuddyError } from "../util/errors";
+import InvocationModel from "./invocation_model";
 
 interface Props {
   model: InvocationModel;
   expanded: boolean;
 }
 
-export default class BuildLogsCardComponent extends React.Component {
-  props: Props;
+interface State {
+  consoleBuffer?: string;
+  nextChunkId?: string;
+  fetchedFirstChunk?: boolean;
+}
+
+const POLL_TAIL_INTERVAL_MS = 1_000;
+const MAX_INITIAL_LINES = 100_000;
+
+const InvocationStatus = invocation.Invocation.InvocationStatus;
+
+export default class BuildLogsCardComponent extends React.Component<Props, State> {
+  state: State = {
+    consoleBuffer: "",
+    nextChunkId: "",
+    fetchedFirstChunk: false,
+  };
+
+  private pollTailTimeout: number | null = null;
+
+  constructor(props: Props) {
+    super(props);
+
+    const invocation = props.model.invocations[0];
+    if (!invocation) {
+      console.error("BuildLogsCard: invocation model is missing invocation");
+    }
+  }
+
+  componentDidMount() {
+    if (this.shouldFetchFromEventLog()) {
+      this.fetchTail(/*isFirstRequest=*/ true);
+    }
+  }
+
+  componentWillUnmount() {
+    window.clearTimeout(this.pollTailTimeout);
+  }
+
+  private shouldFetchFromEventLog() {
+    return capabilities.chunkedEventLogs && Boolean(this.props.model.invocations[0]?.lastChunkId);
+  }
+
+  private isInvocationComplete() {
+    const invocation = this.props.model.invocations[0];
+    return invocation?.invocationStatus === InvocationStatus.COMPLETE_INVOCATION_STATUS;
+  }
+
+  private fetchTail(isFirstRequest = false) {
+    const invocation = this.props.model.invocations[0];
+    let rpcError: BuildBuddyError | null = null;
+    let nextChunkId = "";
+    const wasCompleteBeforeMakingRequest = this.isInvocationComplete();
+
+    rpcService.service
+      .getEventLogChunk(
+        new eventlog.GetEventLogChunkRequest({
+          invocationId: invocation.invocationId,
+          chunkId: this.state.nextChunkId,
+          ...(isFirstRequest && {
+            // For the first request, fetch a large amount of lines starting from
+            // the tail and reading backward.
+            minLines: MAX_INITIAL_LINES,
+            readBackward: true,
+          }),
+        })
+      )
+      .then((response) => {
+        console.log(response);
+        let consoleBuffer = this.state.consoleBuffer;
+        for (const line of response.chunk?.lines || []) {
+          consoleBuffer += String.fromCharCode(...line) + "\n";
+        }
+        nextChunkId = response.nextChunkId;
+        this.setState({ consoleBuffer, nextChunkId });
+      })
+      .catch((e) => {
+        rpcError = BuildBuddyError.parse(e);
+      })
+      .finally(() => {
+        // NotFound / OutOfRange errors just mean the next chunk has not yet been written
+        // and that we should continue polling.
+        if (rpcError?.code === "NotFound" || rpcError?.code === "OutOfRange") {
```

Comment:
When does it return not found and when does it return out of range?

Reply:
IIRC it returns OutOfRange error on the first chunk, although I don't have context as to why. https://github.com/buildbuddy-io/buildbuddy/pull/709/files#diff-53b18e452029fc77ad3cabd57b40b206239e79cec1fc2a6c4ad9208729b2af11R122

@tempoz